app/models/fire_cloud_client.rb

```diff
@@ -1297,26 +1297,16 @@ def get_pet_service_account_key(project_name)
   # * *return*
   #   - Object depends on method, can be one of the following: +Google::Cloud::Storage::Bucket+, +Google::Cloud::Storage::File+,
   #     +Google::Cloud::Storage::FileList+, +Boolean+, +File+, or +String+
-
-  def execute_gcloud_method(method_name, retry_count=0, *params)
+  def execute_gcloud_method(method_name, retry_count = 0, *params)
     begin
       self.send(method_name, *params)
     rescue => e
+      status_code = extract_status_code(e)
       current_retry = retry_count + 1
-      Rails.logger.info "error calling #{method_name} with #{params.join(', ')}; #{e.message} -- attempt ##{current_retry}"
-      retry_time = retry_count * ApiHelpers::RETRY_INTERVAL
-      sleep(retry_time) unless RETRY_BACKOFF_BLACKLIST.include?(method_name)
-      # only retry if status code indicates a possible temporary error, and we are under the retry limit and
-      # not calling a method that is blocked from retries.  In case of a NoMethodError or RuntimeError, use 500 as the
-      # status code since these are unrecoverable errors
-      if e.respond_to?(:code)
-        status_code = e.code
-      elsif e.is_a?(NoMethodError) || e.is_a?(RuntimeError)
-        status_code = 500
-      else
-        status_code = nil
-      end
       if should_retry?(status_code) && retry_count < ApiHelpers::MAX_RETRY_COUNT && !ERROR_IGNORE_LIST.include?(method_name)
+        Rails.logger.info "error calling #{method_name} with #{params.join(', ')}; #{e.message} -- attempt ##{current_retry}"
+        retry_time = retry_count * ApiHelpers::RETRY_INTERVAL
```

Comment:
Could this code notably affect load on SCP or Terra servers?

If so, a more robust `retry_time` might be worthwhile.  The canonical solution is "exponential backoff with jitter", which we've [implemented in Python](https://github.com/broadinstitute/scp-ingest-pipeline/blob/71fc6587bb3fcf19a861d7b6b81ed4693b907a55/ingest/mongo_connection.py#L57) for reference.  That better prevents [thundering herds](https://en.wikipedia.org/wiki/Thundering_herd_problem).

Reply:
Agreed - that approach is better, but the existing retry logic here was written prior to that effort.  This update was only handling the logic around invoking it.  I can take a stab at adding jitter to the retry.

Reply:
And this is why I love Ruby: https://github.com/pawelpacana/exponential-backoff

Reply:
aebcde0